src/boundary_condition.cpp

```diff
@@ -0,0 +1,269 @@
+
+RotationalPeriodicBC::RotationalPeriodicBC(int i_surf, int j_surf)
+  : PeriodicBC(i_surf, j_surf)
+{
+  Surface& surf1 {*model::surfaces[i_surf_]};
+  Surface& surf2 {*model::surfaces[j_surf_]};
+
+  // Check the type of the first surface
+  bool surf1_is_xyplane;
+  if (const auto* ptr = dynamic_cast<const SurfaceXPlane*>(&surf1)) {
+    surf1_is_xyplane = true;
+  } else if (const auto* ptr = dynamic_cast<const SurfaceYPlane*>(&surf1)) {
+    surf1_is_xyplane = true;
+  } else if (const auto* ptr = dynamic_cast<const SurfacePlane*>(&surf1)) {
+    surf1_is_xyplane = false;
+  } else {
+    throw std::invalid_argument(fmt::format("Surface {} is an invalid type for "
+      "rotational periodic BCs. Only x-planes, y-planes, or general planes "
+      "(that are perpendicular to z) are supported for these BCs.", surf1.id_));
+  }
+
+  // Check the type of the second surface
+  bool surf2_is_xyplane;
+  if (const auto* ptr = dynamic_cast<const SurfaceXPlane*>(&surf2)) {
+    surf2_is_xyplane = true;
+  } else if (const auto* ptr = dynamic_cast<const SurfaceYPlane*>(&surf2)) {
+    surf2_is_xyplane = true;
+  } else if (const auto* ptr = dynamic_cast<const SurfacePlane*>(&surf2)) {
+    surf2_is_xyplane = false;
+  } else {
+    throw std::invalid_argument(fmt::format("Surface {} is an invalid type for "
+      "rotational periodic BCs. Only x-planes, y-planes, or general planes "
+      "(that are perpendicular to z) are supported for these BCs.", surf2.id_));
+  }
+
+  // Compute the surface normal vectors and make sure they are perpendicular
+  // to the z-axis
+  Direction norm1 = surf1.normal({0, 0, 0});
+  Direction norm2 = surf2.normal({0, 0, 0});
+  if (std::abs(norm1.z) > FP_PRECISION) {
+    throw std::invalid_argument(fmt::format("Rotational periodic BCs are only "
+      "supported for rotations about the z-axis, but surface {} is not "
+      "perpendicular to the z-axis.", surf1.id_));
+  }
+  if (std::abs(norm2.z) > FP_PRECISION) {
+    throw std::invalid_argument(fmt::format("Rotational periodic BCs are only "
+      "supported for rotations about the z-axis, but surface {} is not "
+      "perpendicular to the z-axis.", surf2.id_));
+  }
+
+  // Make sure both surfaces intersect the origin
+  if (std::abs(surf1.evaluate({0, 0, 0})) > FP_COINCIDENT) {
+    throw std::invalid_argument(fmt::format("Rotational periodic BCs are only "
+      "supported for rotations about the origin, but surface{} does not "
+      "intersect the origin.", surf1.id_));
+  }
+  if (std::abs(surf2.evaluate({0, 0, 0})) > FP_COINCIDENT) {
+    throw std::invalid_argument(fmt::format("Rotational periodic BCs are only "
+      "supported for rotations about the origin, but surface{} does not "
+      "intersect the origin.", surf2.id_));
+  }
+
+  // Compute the angle between the two surfaces; this is the BC rotation angle
+  double theta1 = std::atan2(norm1.y, norm1.x);
+  double theta2 = std::atan2(norm2.y, norm2.x);
+  angle_ = theta2 - theta1;
```

Comment:
It looks like this doesn't work for all cases. I tried doing a 1/6th rotational symmetry problem with the following:
```Python
plane1 = openmc.YPlane(boundary_type='periodic')

angle = pi/3
p1 = (0., 0., 0.)
p2 = (cos(angle), sin(angle), 0.)
p3 = (0., 0., 1.)
plane2 = openmc.Plane.from_points(p1, p2, p3, boundary_type='periodic')
plane2.periodic_surface = plane1
```
Note that for plane 2, the normal is (p2 - p1) x (p3 - p1) which results in a direction that is down and to the right. Here is what you end up with (planes in green, normals in blue):
<img width="282" alt="wrong_angle" src="https://user-images.githubusercontent.com/743095/96718271-61511980-136d-11eb-9b0e-845a20ac9567.png">
As you can see, we end up with an angle between the normals that is 2π/3 instead of π/3 as desired. The end result is that when you try to run a model like this, all particles that cross the periodic boundary get lost. If you swap `p2` and `p3` in the definition of the plane, the normal will go in the other direction, the calculated angle is π/3, and everything works as expected.

Reply:
Hmm, thinking about this a little more, if instead of switching the planes, I use the region below the y-plane (which gives me 1/3rd rotational symmetry), the simulation works fine because in that case the calculated angle matches the region. So this implies that how the angle is calculated depends on how the surfaces are actually used to create a cell, which seems awkward. Maybe we need some sort of convention for how a model must be constructed if rotational periodic BCs are used?

Reply:
Yeah, I haven't figured out a good solution to that problem yet. When I wrote this code, I was operating off the assumption that one normal vector would normally point inward, and the other outward. (This is likely what happens if you have an azimuthally segmented problem and you define all the planes from the general `Plane` class in a loop.) But this assumption seems to be false often enough.

Perhaps we'll have to specify a convention, but I'll keep thinking to see if we can come up with something easy. Maybe we can compute the bounding box of the root universe and check it against the normals of the bounding planes.

Reply:
Sounds good -- let me know if you come up with something clever. The bounding box idea sounds a little complicated, so if it's too difficult we can always just adopt some convention that is expected to be obeyed. I think as long as it is well documented, then at least we have something we can point to if (and when) a user messes it up.